File: interpreter.py
#!/usr/bin/python3

import re
import argparse
import os.path


verbose = False

def init_vm(memspace, consts):
    print_log("init vm...")
    print_log("memspace: %d" % memspace)
    mem = [None] * memspace
    index = memspace - 1
    while consts:
        const = consts.pop(0)
        mem[index] = const
        index = index - 1
    print_log("mem: %s" % mem)
    print_log("init done...")
    return mem

def init_labels(commands):
    labels = {}
    p = re.compile("(?P<label>\w+):")
    for index,cmd in enumerate(commands):
        m = p.match(cmd)
        if m:
            lbl = m.group("label")
            labels[lbl] = index
    print_log("labels: %s" % labels)
    print_log("init labels done...")
    return labels
        
def run_commands(mem, commands, inputs):
    print_log("start running commands...")
    labels = init_labels(commands)
    outputs = interpret(mem, labels, commands, inputs)
    print_log("end running commands...")
    return outputs

def interpret(mem, labels, commands, inputs):
    outputs = []
    steps = 0
    print_log("interpreting...")
    p = re.compile("\[(?P<addr>\w+)\]")
    sub = re.compile("(\s)*SUB(\s)+\[(?P<addr>\w+)\]|(\s)*SUB(\s)+((?P<index>\w+))")
    add = re.compile("(\s)*ADD(\s)+\[(?P<addr>\w+)\]|(\s)*ADD(\s)+((?P<index>\w+))")
    jpz = re.compile("(\s)*JUMPZ(\s)+(?P<label>\w+)")
    jpn = re.compile("(\s)*JUMPN(\s)+(?P<label>\w+)")
    jmp = re.compile("(\s)*JUMP(\s)+(?P<label>\w+)")
    bup = re.compile("(\s)*BUMPUP(\s)+\[(?P<addr>\w+)\]|(\s)*BUMPUP(\s)+(?P<index>\w+)")
    bdn = re.compile("(\s)*BUMPDN(\s)+\[(?P<addr>\w+)\]|(\s)*BUMPDN(\s)+(?P<index>\w+)")
    cpt = re.compile("(\s)*COPYTO(\s)+\[(?P<addr>\w+)\]|(\s)*COPYTO(\s)+(?P<index>\w+)")
    cpf = re.compile("(\s)*COPYFROM(\s)+\[(?P<addr>\w+)\]|(\s)*COPYFROM(\s)+(?P<index>\w+)")
    ptr = 0
    x = None
    while True:
        print_log("mem: %s" % mem)
        cmd = commands[ptr]
        print_log("interpreting command: %s" % cmd)
        if "INBOX" in cmd:
            if not inputs:
                break
            x = inputs.pop(0)
        elif "OUTBOX" in cmd:
            outputs.append(x)
            x = None
        elif "ADD" in cmd:
            m = add.match(cmd)
            if m.group("addr"):
                i = get_addr(m)
                addr = get_val_from_mem(mem, i)
                x = x + mem[addr]
            else:
                i = get_index(m)
                x = x + mem[i]
        elif "SUB" in cmd:
            m = sub.match(cmd)
            if m.group("addr"):
                i = get_addr(m)
                addr = get_val_from_mem(mem, i)
                x = x - mem[addr]
            else:
                i = get_index(m)
                x = x - mem[i]
        elif "JUMPZ" in cmd:
            if x == 0:
                m = jpz.match(cmd)
                lbl = m.group("label")
                ptr = labels[lbl]
        elif "JUMPN" in cmd:
            if x < 0:
                m = jpn.match(cmd)
                lbl = m.group("label")
                ptr = labels[lbl]
        elif "JUMP" in cmd:
            m = jmp.match(cmd)
            lbl = m.group("label")
            ptr = labels[lbl]
        elif "BUMPUP" in cmd:
            m = bup.match(cmd)
            if m.group("addr"):
                i = get_addr(m)
                addr = get_val_from_mem(mem, i)
                mem[addr] = mem[addr] + 1
                x = mem[addr]
            else:
                i = get_index(m)
                mem[i] = mem[i] + 1
                x = get_val_from_mem(mem, i)
        elif "BUMPDN" in cmd:
            m = bdn.match(cmd)
            if m.group("addr"):
                i = get_addr(m)
                addr = get_val_from_mem(mem, i)
                mem[addr] = mem[addr] - 1
                x = mem[addr]
            else:
                i = get_index(m)
                mem[i] = mem[i] - 1
                x = get_val_from_mem(mem, i)
        elif "COPYTO" in cmd:
            m = cpt.match(cmd)
            if m.group("addr"):
                i = get_addr(m)
                addr = get_val_from_mem(mem, i)
                mem[addr] = x
            else:
                i = get_index(m)
                mem[i] = x
        elif "COPYFROM" in cmd:
            m = cpf.match(cmd)
            if m.group("addr"):
                i = get_addr(m)
                addr = get_val_from_mem(mem, i)
                x = mem[addr]
            else:
                i = get_index(m)
                x = get_val_from_mem(mem, i)
        elif "END" in cmd:
            break
        ptr = ptr + 1
        steps = steps + 1
    print_log("interpreted in %d steps with %d commands" % (steps, len(commands)))
    return outputs

def get_val_from_mem(mem, i):
    val = mem[i]
    return get_raw_val(val)

def get_index(m):
    return int(m.group("index"))

def get_addr(m):
    return int(m.group("addr"))

def get_raw_val(val):
    if val is not None :
        try:
            return int(val)
        except ValueError:
            if val.isalpha():
                return ord(val)
            return int(val)
    return None

def get_display_val(val):
    if val is not None:
        try:
            if chr(val).isalpha():
                return chr(val)
            else:
                return ord(chr(val))
        except ValueError:
            return val
    return None

def read_commands(file_name):
    commands = []
    with open(file_name, "r") as f:
        commands = f.readlines()

    f.close()

    commands = [cmd.replace("\t", " ") for cmd in commands]
    return commands

def read_inputs(file_name):
    line = ""
    with open(file_name, "r") as f:
        line = f.readline()
    f.close()

    inputs = [get_raw_val(val) for val in line.strip().split(" ")]
    return inputs

def read_init(file_name):
    lines = []
    memspace = None
    constants = []
    with open(file_name, "r") as f:
        lines = f.readlines()

    f.close()

    memspace = int(lines[0].strip())
    memspace = get_raw_val(lines[0].strip())
    if len(lines) == 2:
        constants = [get_raw_val(val) for val in lines[1].strip().split(" ")]
    return memspace, constants

def print_log(line):
    if verbose:
        print(line)

def print_error(line):
    print("Error: %s" % line)

def check_file(filename):
    if not os.path.isfile(filename):
        print_error("%s not found" % filename)
        exit(1)

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description="HRM Instructions Interpreter")
    parser.add_argument("--init", help="file path to init VM data", default="init_vm.txt", required=True, dest="init_filename")
    parser.add_argument("--cmd", help="file path to commands data", default="commands.txt", required=True, dest="cmd_filename")
    parser.add_argument("--input", help="file path to input data", default="inputs.txt", required=True, dest="in_filename")
    parser.add_argument("--verbose", "-v", help="print log", action="count")
    args = parser.parse_args()
    init_filename = args.init_filename
    check_file(init_filename)
    cmd_filename = args.cmd_filename
    check_file(cmd_filename)
    in_filename = args.in_filename
    check_file(in_filename)
    verbose = True if args.verbose else False
    memspace, constants = read_init(init_filename)
    mem = init_vm(memspace, constants)
    commands = read_commands(cmd_filename)
    inputs = read_inputs(in_filename)
    print("inputs: %s" % [get_display_val(val) for val in inputs])
    outputs = run_commands(mem, commands, inputs)
    print("outputs: %s" % [get_display_val(val) for val in outputs])

File: interpreter.rb
#!/usr/bin/ruby

require 'optparse'

class HRMInterpreter
  attr_accessor :verbose
  attr_reader :memspace, :mem, :constants, :commands, :inputs, :labels,
              :outputs

  def initialize(parser)
    @verbose = parser.verbose
    @memspace = 0
    @mem = []
    @constants = []
    @commands = []
    @inputs = []
    @labels = {}
    @outputs = []
    read_init(parser.init_filename)
    read_commands(parser.cmd_filename)
    read_inputs(parser.in_filename)
    init_vm
    init_labels
  end

  def init_vm
    print_log("init vm...")
    print_log(format("memspace: %d", @memspace))
    @mem = [nil] * @memspace
    index = @memspace - 1
    @constants.each do |const| 
      @mem[index] = const
      index = index - 1
    end
    print_log(format("mem: %s", @mem))
    print_log("init done...")
  end

  def init_labels
    @commands.each_with_index do |cmd,index|
      m = cmd.match(/(?<label>\w+):/)
      if !m.nil?
        lbl = m['label']
        @labels[lbl] = index  
      end
    end
    print_log(format("labels: %s", labels))
    print_log("init labels done...")
  end
        
  def read_init(file_name)
    print_log("read_init...")
    lines = []
    lines = File.readlines(file_name)
    @memspace = lines.first.strip.to_i
    @memspace = get_raw_val(lines.first.strip)
    if lines.length == 2
      lines[1].strip.split(" ").each { |val| @constants.push(get_raw_val(val)) }
    end
  end

  def read_commands(file_name)
    print_log("read_commands...")
    @commands = File.readlines(file_name)
    @commands.each { |cmd| cmd.strip! }
  end

  def read_inputs(file_name)
    print_log("read_inputs...")
    line = ""
    line = File.read(file_name)
    line.strip.split(" ").each { |val| @inputs.push(get_raw_val(val)) }
  end

  def add_raw(left, right)
    if left.instance_of?(String) && right.instance_of?(String)
      print_error(format("Unable to add values of non-integer types"))
      exit
    end
    left + right
  end

  def sub_raw(left, right)
    result = 0
    if left.class != right.class
      print_error(format("Unable to sub values of different types %s, %s",
                          left.class, right.class))
      exit
    elsif left.instance_of?(Fixnum) && right.instance_of?(Fixnum)
      result = left - right
    elsif left.instance_of?(String) && right.instance_of?(String)
      lm = left.match(/[^\d]+/)
      rm = right.match(/[^\d]+/)
      if lm.nil? and rm.nil?
        result = left.to_i - right.to_i
      elsif !lm.nil? and !rm.nil?
        result = left.ord - right.ord
      else
        print_error(format("Unable to sub values of different types %s, %s",
                            left.class, right.class))
        exit
      end
    end
    result
  end

  def cmp_raw(left, op, right)
    if left.instance_of?(String)
      lm = left.match(/[^\d]+/)
      left = lm.nil? ? left.to_i : left.ord
    end
    if right.instance_of?(String)
      rm = right.match(/[^\d]+/)
      right = rm.nil? ? right.to_i : right.ord
    end
    compare(left, op, right)
  end

  def compare(left, op, right)
    result =
      case op
      when "eq" then left == right
      when "ne" then left != right
      when "lt" then left < right
      when "gt" then left > right
      when "lte" then left <= right
      when "gte" then left >= right
      else
        print_error(format("Invalid operation %s", op))
        exit
      end
    result
  end

  def get_val_from_mem(i)
    i = i.to_i if i.instance_of?(String)
    @mem[i]
  end

  def set_val_to_mem(i, val)
    i = i.to_i if i.instance_of?(String)
    @mem[i] = val
  end

  def get_raw_val(val)
    raw_val = if !val.nil?
                !val.match(/\d+/).nil? ? val.to_i : val
              else
                nil
              end
    raw_val
  end

  def print_log(line)
    puts line if verbose
  end

  def print_error(line)
    puts format("Error: %s", line)
  end

  def print_warn(line)
    puts format("Warning: %s", line)
  end

  def to_s
    print @mem
    puts
    print @memspace
    puts
    print @constants
    puts
    print @inputs
    puts
    print @commands
    puts
    print @labels
    puts
  end
end

class HRMOptionParser
  attr_accessor :init_filename, :cmd_filename, :in_filename, :verbose

  def initialize
    @init_filename = nil
    @cmd_filename = nil
    @in_filename = nil
    @verbose = false
  end

  def parse!
    parser = OptionParser.new do |opts|
      opts.banner = "Usage: interpreter.rb [options]"

      opts.on('-n', '--init init_filename', 'Init filename') do |init_filename|
        @init_filename = init_filename
      end

      opts.on('-c', '--cmd cmd_filename', 'Commands filename') do |cmd_filename|
        @cmd_filename = cmd_filename
      end

      opts.on('-i', '--in in_filename', 'Inputs filename') do |in_filename|
        @in_filename = in_filename
      end

      opts.on('-v', '--verbose', 'Verbose') do
        @verbose = true
      end

      opts.on('-h', '--help', 'Help') do
        puts opts
        exit
      end
    end

    parser.parse!

    check_required

    check_file
  end

  def check_required
    required = []
    if @init_filename.nil?
      required.push("--init")
    end
    if @cmd_filename.nil?
      required.push("--cmd")
    end
    if @in_filename.nil?
      required.push("--in")
    end
    if !required.empty?
      puts "error: the following arguments are required: #{required.join(', ')}"
      exit
    end
  end

  def check_file
    [@init_filename, @cmd_filename, @in_filename].each do |filename|
      if !File.exist?(filename)
        puts format("%s not found", filename)
        exit
      end
    end
  end
end

if __FILE__ == $0
  parser = HRMOptionParser.new
  parser.parse!
  interpreter = HRMInterpreter.new(parser)

  print interpreter.to_s
end 
